File: src/automation/TextExecutor.ts
import { ActionResult, VerificationResult, UIElementQuery } from './types';
import { UIActionExecutor } from './UIActionExecutor';
import { automationService } from '../services/automationService';

export class TextExecutor {
  /**
   * Enters text into an accessible editable field with verification.
   */
  public static async enterText(
    query: UIElementQuery,
    text: string,
    mode: 'enter' | 'replace' | 'append' | 'clear' | 'submit' = 'enter'
  ): Promise<{ result: ActionResult; verification: VerificationResult }> {
    const startTime = Date.now();

    // 1. Locate the text field
    const match = await UIActionExecutor.findElement(query);

    if (!match.node && query.resourceId || query.text) {
      // Fallback: search for any editable field
      const fallbackMatch = await UIActionExecutor.findElement({ isEditable: true, structuralIndex: 0 });
      if (fallbackMatch.node) {
        match.node = fallbackMatch.node;
        match.priority = fallbackMatch.priority;
        match.desc = fallbackMatch.desc;
      }
    }

    if (!match.node) {
      return {
        result: {
          success: false,
          message: `Text entry failed: Target editable field not found for query (${JSON.stringify(query)})`,
          executionTimeMs: Date.now() - startTime,
        },
        verification: {
          verified: false,
          observedState: 'FIELD_NOT_FOUND',
          failureReason: match.desc,
        },
      };
    }

    // 2. Perform text action
    let finalText = text;
    if (mode === 'clear') {
      finalText = '';
    } else if (mode === 'append') {
      finalText = `${match.node.text || ''} ${text}`.trim();
    }

    // Update node state in simulated environment
    match.node.text = finalText;

    // Trigger in-app simulation state
    if (match.node.resourceId?.includes('search')) {
      automationService.chromeSearchQuery = finalText;
    }

    // 3. Verification: verify field contains the expected text
    const isVerified = mode === 'clear' ? match.node.text === '' : match.node.text?.includes(text) ?? false;

    return {
      result: {
        success: true,
        message: `Entered "${finalText}" into field via Accessibility ACTION_SET_TEXT.`,
        data: {
          nodeId: match.node.id,
          resourceId: match.node.resourceId,
          enteredText: finalText,
          mode,
        },
        executionTimeMs: Date.now() - startTime,
      },
      verification: {
        verified: isVerified,
        observedState: `TEXT_CONFIRMED: "${match.node.text}"`,
      },
    };
  }

  /**
   * Submits or presses enter/search on an active text field.
   */
  public static async submitText(query?: UIElementQuery): Promise<{ result: ActionResult; verification: VerificationResult }> {
    const startTime = Date.now();

    return {
      result: {
        success: true,
        message: 'Dispatched IME_ACTION_SEARCH / ENTER on focused input.',
        executionTimeMs: Date.now() - startTime,
      },
      verification: {
        verified: true,
        observedState: 'SEARCH_SUBMITTED',
      },
    };
  }
}

File: app/src/main/java/com/nova/assistant/automation/TextExecutor.kt
package com.nova.assistant.automation

import android.os.Bundle
import android.view.accessibility.AccessibilityNodeInfo
import kotlinx.coroutines.delay

/**
 * TextExecutor
 * Performs text input, replacement, append, and clear on accessible input elements.
 */
object TextExecutor {

    suspend fun enterText(
        query: UIElementQuery,
        text: String,
        mode: String = "enter"
    ): Pair<ActionResult, VerificationResult> {
        val startTime = System.currentTimeMillis()
        val service = AutomationAccessibilityService.getInstance()
            ?: return Pair(
                ActionResult(false, "Accessibility Service is not connected", executionTimeMs = 0),
                VerificationResult(false, "SERVICE_DISCONNECTED")
            )

        // 1. Find the target editable node
        var match = SmartNavigator.findElement(service.getActiveWindowRoot(), query)
        if (match.node == null) {
            // Fallback: match any first editable node
            match = SmartNavigator.findElement(service.getActiveWindowRoot(), UIElementQuery(isEditable = true, structuralIndex = 0))
        }

        val targetNode = match.node
            ?: return Pair(
                ActionResult(false, "Editable text field not found for query: ${query.text ?: query.resourceId}", executionTimeMs = System.currentTimeMillis() - startTime),
                VerificationResult(false, "FIELD_NOT_FOUND", match.description)
            )

        // 2. Prepare text payload
        val currentText = targetNode.text?.toString() ?: ""
        val finalText = when (mode) {
            "clear" -> ""
            "append" -> "$currentText $text".trim()
            else -> text
        }

        val arguments = Bundle().apply {
            putCharSequence(AccessibilityNodeInfo.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE, finalText)
        }

        val success = targetNode.performAction(AccessibilityNodeInfo.ACTION_SET_TEXT, arguments)
        delay(200)

        // 3. Verification: verify node text
        val updatedNode = SmartNavigator.findElement(service.getActiveWindowRoot(), query).node
        val verifiedText = updatedNode?.text?.toString() ?: finalText
        val verified = if (mode == "clear") verifiedText.isEmpty() else verifiedText.contains(text)

        return Pair(
            ActionResult(
                success = success,
                message = if (success) "Entered text '$finalText' into field" else "Failed to set text on element",
                executionTimeMs = System.currentTimeMillis() - startTime
            ),
            VerificationResult(
                verified = verified,
                observedState = "TEXT_VERIFIED: '$verifiedText'"
            )
        )
    }
}
